### PR TITLE
Always show use of unassigned var errors

### LINKED ISSUE #15298: Call to local function referencing then-undeclared variable incorrectly allowed
**Version Used**: 2.0.0.61104 - which is in the recent VS2017RC release.

(Don't have a dev build set up, so I apologize if this has been fixed already - scanned through local function issues and couldn't find any related items, and didn't quickly see any tests with this specific case)

**Repro code**:

```csharp
class Program
{
    static void Main(string[] args)
    {
        Foo();
        int x = 2;
        void Foo() => System.Console.WriteLine(x); // prints "0"
    }
}
```

**Expected Behavior**:

Something similar to the following example:

```csharp
class Program
{
    static void Main(string[] args)
    {
        int x;
        Foo(); // CS0165: Use of unassigned local variable 'x'
        x = 2;
        void Foo() => System.Console.WriteLine(x);
    }
}
```

Ping @dotnet/roslyn-compiler - probably @agocke in particular?

### LINKED ISSUE #15322: Roslyn fails to enforce definitely assigned rules and cannot emit code for local functions with closures over pattern variables in switch statements
Following code should have error CS0165: Use of unassigned local variable 'x'

```cs
class Program
{
    static void Main(object[] args)
    {
        switch(args[0])
        {
            case string x:
                Foo(); // missing error
                break;
            case int x:
                void Foo() => System.Console.WriteLine(x);
                break;
        }
    }
}
```

Roslyn doesn't detect the error and fails to compile the code, this is VS2017 RC output:
```
csc.exe exited with code 255
```

Note that if you reorder the cases you get correct error i.e.
```cs
class Program
{
    static void Main(object[] args)
    {
        switch(args[0])
        {
            case int x:
                void Foo() => System.Console.WriteLine(x);
                break;
            case string x:
                Foo(); // error CS0165: Use of unassigned local variable 'x'
                break;
        }
    }
}
```

TryRoslyn emits following stacktrace for original code (it may be a bit outdated though):
```
The given key was not present in the dictionary. 
at System.Collections.Generic.Dictionary`2.get_Item(TKey key) 
at Microsoft.CodeAnalysis.CodeGen.LocalSlotManager.GetLocal(ILocalSymbol symbol) in ...\Src\Compilers\Core\Portable\CodeGen\LocalSlotManager.cs:line 148 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.GetLocal(LocalSymbol symbol) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitStatement.cs:line 1387 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.GetLocal(BoundLocal localExpression) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitStatement.cs:line 1382 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitLocalAddress(BoundLocal localAccess) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitAddress.cs:line 167 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitAddress(BoundExpression expression, AddressKind addressKind) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitAddress.cs:line 38 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitArgument(BoundExpression argument, RefKind refKind) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitExpression.cs:line 594 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitArguments(ImmutableArray`1 arguments, ImmutableArray`1 parameters) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitExpression.cs:line 769 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitCallExpression(BoundCall call, UseKind useKind) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitExpression.cs:line 1480 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitExpressionCore(BoundExpression expression, Boolean used) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitExpression.cs:line 92 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitExpressionCoreWithStackGuard(BoundExpression expression, Boolean used) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitExpression.cs:line 72 
at Microsoft.CodeAnalysis.CSharp.CodeGen.CodeGenerator.EmitExpression(BoundExpression expression, Boolean used) in ...\Src\Compilers\CSharp\Portable\CodeGen\EmitExpression.cs:line 60 
```

ping @agocke

### PR DESCRIPTION
**Pull request by agocke**
Prior to C# 7 we tried to avoid giving duplicate errors
when the user used a variable before it was defined (which
is both a use-before-defined and a use-before-assigned error),
but the line here is blurred with local functions, which can
use variables before they are declared in subtle ways that can't
be detected based on declaration order in syntax.

It may be possible to provide the earlier behavior, but the
current optimization is definitely insufficient and hides
legitimate errors in user programs.

Fixes #15322
Fixes #15298

### CODE REVIEW

**Comment by agocke** [comment]
\+ @gafter  @dotnet/roslyn-compiler For review

**Comment by genlu** [comment]
NuGet restore failed https://ci.dot.net/job/dotnet_roslyn/job/master/job/linux_debug_prtest/447/

@dotnet-bot retest linux_debug_prtest  please

**Comment by agocke** [comment]
Hmm, there are a lot more additional errors in existing cases than I expected. I'll see if there's a way I can preserve the existing behavior for non-local functions.

**Comment by agocke** [comment]
@dotnet/roslyn-compiler This is ready for review

**Comment by jaredpar** [comment]
This is a breaking change correct?

**Comment by agocke** [comment]
@jaredpar With the latest change, no I believe it should only affect local function error reporting.

However, even with the original change it could only result in *more* errors, not breaking code that compiles.

File: src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs [hunk 1]
```diff
@@ -896,37 +896,15 @@ protected void CheckAssigned(Symbol symbol, SyntaxNode node)
...
-        private void ReportUnassigned(Symbol symbol, SyntaxNode node, int? slotOpt)
+        /// <param name="symbol">Symbol to variable that is unassigned.</param>
+        /// <param name="node">Syntax where read occurs.</param>
+        /// <param name="slotOpt">Optional slot where variable is located.</param>
+        /// <param name="considerDefineLocation">
+        /// True if error reporting should consider the location where the
+        /// variable is defined (for instance, eliding errors about reading
+        /// variables that have not yet been defined).
+        /// </param>
+        private void ReportUnassigned(Symbol symbol, SyntaxNode node, int? slotOpt, bool considerDefineLocation = true)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs]: >considerDefineLocation [](start = 89, length = 22)

considerDeclarationLocation? Actually, consider using even more meaningful name instead, something like "doNotReportUseBeforeDeclaration" or similar.

File: src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs [hunk 2]
```diff
@@ -61,7 +61,41 @@ private void ReplayReads(ref BitVector reads, SyntaxNode syntax)
...
+                {
+                    if (slot >= this.State.Assigned.Capacity)
+                    {
+                        Normalize(ref this.State);
+                    }
+
+                    if (slot > 0 && !this.State.IsAssigned(slot))
+                    {
+                        // Local functions can "call forward" to after a variable has
+                        // been defined but before it has been assigned, so we can never
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs]: >defined [](start = 32, length = 7)

declared?

File: src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs [hunk 3]
```diff
@@ -61,7 +61,41 @@ private void ReplayReads(ref BitVector reads, SyntaxNode syntax)
...
+                        // consider the definition location when reporting errors.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs]: >definition [](start = 40, length = 10)

declaration?

**Comment by AlekseyTs** [comment]
LGTM

**Comment by agocke** [comment]
@dotnet/roslyn-compiler Ping for one more review

**Comment by VSadov** [comment]
LGTM

File: src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs [hunk 4]
```diff
@@ -61,7 +61,41 @@ private void ReplayReads(ref BitVector reads, SyntaxNode syntax)
...
+        private void CheckIfAssignedDuringLocalFunctionReplay(Symbol symbol, SyntaxNode node, int slot)
+        {
+            Debug.Assert(!IsConditionalState);
+            if ((object)symbol != null)
+            {
+                NoteRead(symbol);
+
+                if (this.State.Reachable)
+                {
+                    if (slot >= this.State.Assigned.Capacity)
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs]: This shouldn't have any effect on the behavior. If it does, something may be wrong (possibly my understanding).

**Pull request merged**